src/render/draw_symbol.js

```diff
@@ -116,35 +121,45 @@ function updateVariableAnchors(coords, painter, layer, sourceCache, rotationAlig
     const tr = painter.transform;
     const rotateWithMap = rotationAlignment === 'map';
     const pitchWithMap = pitchAlignment === 'map';
+    const tileTransform = tr.projection.createTileTransform(tr, tr.worldSize);
 
     for (const coord of coords) {
         const tile = sourceCache.getTile(coord);
         const bucket: SymbolBucket = (tile.getBucket(layer): any);
-        if (!bucket || !bucket.text || !bucket.text.segments.get().length) continue;
+        if (!bucket || bucket.projection !== tr.projection.name || !bucket.text || !bucket.text.segments.get().length) {
+            continue;
+        }
 
         const sizeData = bucket.textSizeData;
         const size = symbolSize.evaluateSizeForZoom(sizeData, tr.zoom);
 
         const pixelsToTileUnits = painter.transform.calculatePixelsToTileUnitsMatrix(tile);
-        const labelPlaneMatrix = symbolProjection.getLabelPlaneMatrix(coord.projMatrix, pitchWithMap, rotateWithMap, painter.transform, pixelsToTileUnits);
+        const labelPlaneMatrix = symbolProjection.getLabelPlaneMatrix(coord.projMatrix, tile.tileID.canonical, pitchWithMap, rotateWithMap, painter.transform, pixelsToTileUnits);
         const updateTextFitIcon = layer.layout.get('icon-text-fit') !== 'none' &&  bucket.hasIconData();
 
         if (size) {
             const tileScale = Math.pow(2, tr.zoom - tile.tileID.overscaledZ);
-            const elevation = tr.elevation;
-            const getElevation = elevation ? (p => elevation.getAtTileOffset(coord, p.x, p.y)) : (_ => 0);
             updateVariableAnchorsForBucket(bucket, rotateWithMap, pitchWithMap, variableOffsets, symbolSize,
-                                  tr, labelPlaneMatrix, coord.projMatrix, tileScale, size, updateTextFitIcon, getElevation);
+                                  tr, labelPlaneMatrix, coord, tileScale, size, updateTextFitIcon, tileTransform);
         }
     }
 }
 
 function updateVariableAnchorsForBucket(bucket, rotateWithMap, pitchWithMap, variableOffsets, symbolSize,
-                               transform, labelPlaneMatrix, posMatrix, tileScale, size, updateTextFitIcon, getElevation) {
+                               transform, labelPlaneMatrix, coord, tileScale, size, updateTextFitIcon, tileTransform) {
     const placedSymbols = bucket.text.placedSymbolArray;
     const dynamicTextLayoutVertexArray = bucket.text.dynamicLayoutVertexArray;
     const dynamicIconLayoutVertexArray = bucket.icon.dynamicLayoutVertexArray;
     const placedTextShifts = {};
+    const projMatrix = coord.projMatrix;
+    const elevation = transform.elevation;
+    const getElevation = elevation ? (p => {
+        const e = elevation.getAtTileOffset(coord, p.x, p.y);
+        const up = tileTransform.upVector(coord.canonical, p.x, p.y);
+        const upScale = tileTransform.upVectorScale(coord.canonical);
+        vec3.scale(up, up, e * upScale);
+        return up;
+    }) : (_ => [0, 0, 0]);
```

Comment:
TODO: Share with other references [here](https://github.com/mapbox/mapbox-gl-js/pull/11329/files?authenticity_token=RcIfSnVWLHFxEurquEvhcSNJ%2BYeQSsCZqS9Hx7wHssA5Ol6KIGTI%2BgYdyIjDsZIS%2FAi0MwWfovx4k20CqY%2FUag%3D%3D&file-filters%5B%5D=.glsl&file-filters%5B%5D=.js&file-filters%5B%5D=.json#diff-32491b8d513a9d3eb4a65cd5b5ded4b70be4f2cea814adbc54f437eaccf1af0cR339) and [here](https://github.com/mapbox/mapbox-gl-js/pull/11329/files?authenticity_token=RcIfSnVWLHFxEurquEvhcSNJ%2BYeQSsCZqS9Hx7wHssA5Ol6KIGTI%2BgYdyIjDsZIS%2FAi0MwWfovx4k20CqY%2FUag%3D%3D&file-filters%5B%5D=.glsl&file-filters%5B%5D=.js&file-filters%5B%5D=.json#diff-a3bf6fa30c8d788938050bd6af5d8579ead6636a25f0b719b1ec1b909e4a80f9R140) of `getElevation`

Reply:
https://github.com/mapbox/mapbox-gl-js/pull/11329/commits/ebe30b34ac8e760e72572ce1e0ff79d3a6116939